README.md
# Auto-PyTorch

Copyright (C) 2021  [AutoML Groups Freiburg and Hannover](http://www.automl.org/)

While early AutoML frameworks focused on optimizing traditional ML pipelines and their hyperparameters, another trend in AutoML is to focus on neural architecture search. To bring the best of these two worlds together, we developed **Auto-PyTorch**, which jointly and robustly optimizes the network architecture and the training hyperparameters to enable fully automated deep learning (AutoDL).

Auto-PyTorch is mainly developed to support tabular data (classification, regression) and time series data (forecasting).
The newest features in Auto-PyTorch for tabular data are described in the paper ["Auto-PyTorch Tabular: Multi-Fidelity MetaLearning for Efficient and Robust AutoDL"](https://arxiv.org/abs/2006.13799) (see below for bibtex ref).
Details about Auto-PyTorch for multi-horizontal time series forecasting tasks can be found in the paper ["Efficient Automated Deep Learning for Time Series Forecasting"](https://arxiv.org/abs/2205.05511) (also see below for bibtex ref).

Also, find the documentation [here](https://automl.github.io/Auto-PyTorch/master).


***From v0.1.0, AutoPyTorch has been updated to further improve usability, robustness and efficiency by using SMAC as the underlying optimization package as well as changing the code structure. Therefore, moving from v0.0.2 to v0.1.0 will break compatibility. 
In case you would like to use the old API, you can find it at [`master_old`](https://github.com/automl/Auto-PyTorch/tree/master-old).***

## Workflow

The rough description of the workflow of Auto-Pytorch is drawn in the following figure.

![AutoPyTorch Workflow](https://raw.githubusercontent.com/automl/Auto-PyTorch/master/figs/apt_workflow.png)

In the figure, **Data** is provided by user and
**Portfolio** is a set of configurations of neural networks that work well on diverse datasets.
The current version only supports the *greedy portfolio* as described in the paper *Auto-PyTorch Tabular: Multi-Fidelity MetaLearning for Efficient and Robust AutoDL*
This portfolio is used to warm-start the optimization of SMAC.
In other words, we evaluate the portfolio on a provided data as initial configurations.
Then API starts the following procedures:
1. **Validate input data**: Process each data type, e.g. encoding categorical data, so that Auto-Pytorch can handled.
2. **Create dataset**: Create a dataset that can be handled in this API with a choice of cross validation or holdout splits.
3. **Evaluate baselines** 
   * ***Tabular dataset*** *1: Train each algorithm in the predefined pool with a fixed hyperparameter configuration and dummy model from `sklearn.dummy` that represents the worst possible performance.
   * ***Time Series Forecasting dataset*** : Train a dummy predictor that repeats the last observed value in each series
4. **Search by [SMAC](https://github.com/automl/SMAC3)**:\
    a. Determine budget and cut-off rules by [Hyperband](https://jmlr.org/papers/volume18/16-558/16-558.pdf)\
    b. Sample a pipeline hyperparameter configuration *2 by SMAC\
    c. Update the observations by obtained results\
    d. Repeat a. -- c. until the budget runs out
5. Build the best ensemble for the provided dataset from the observations and [model selection of the ensemble](https://www.cs.cornell.edu/~caruana/ctp/ct.papers/caruana.icml04.icdm06long.pdf).

*1: Baselines are a predefined pool of machine learning algorithms, e.g. LightGBM and support vector machine, to solve either regression or classification task on the provided dataset

*2: A pipeline hyperparameter configuration specifies the choice of components, e.g. target algorithm, the shape of neural networks, in each step and 
(which specifies the choice of components in each step and their corresponding hyperparameters.

## Installation
Clone the repository with submodules 
```sh

git clone --recurse-submodules git@github.com:automl/Auto-PyTorch.git

```

### PyPI Installation

```sh

pip install autoPyTorch

```

Auto-PyTorch for Time Series Forecasting requires additional dependencies 

```sh

pip install 'autoPyTorch[forecasting]'

```

### Manual Installation

We recommend using Anaconda for developing as follows:

```sh
# Following commands assume the user is in a cloned directory of Auto-Pytorch

# We also need to initialize the automl_common repository as follows
# You can find more information about this here:
# https://github.com/automl/automl_common/
git submodule update --init --recursive

# Create the environment
conda create -n auto-pytorch python=3.8
conda activate auto-pytorch
conda install swig
python setup.py install

```

Similarly, to install all the dependencies for Auto-PyTorch-TimeSeriesForecasting:


```sh

git submodule update --init --recursive

conda create -n auto-pytorch python=3.8
conda activate auto-pytorch
conda install swig
pip install -e[forecasting]

```

## Examples

In a nutshell:

```py
from autoPyTorch.api.tabular_classification import TabularClassificationTask

# data and metric imports
import sklearn.model_selection
import sklearn.datasets
import sklearn.metrics
X, y = sklearn.datasets.load_digits(return_X_y=True)
X_train, X_test, y_train, y_test = \
        sklearn.model_selection.train_test_split(X, y, random_state=1)

# initialise Auto-PyTorch api
api = TabularClassificationTask()

# Search for an ensemble of machine learning algorithms
api.search(
    X_train=X_train,
    y_train=y_train,
    X_test=X_test,
    y_test=y_test,
    optimize_metric='accuracy',
    total_walltime_limit=300,
    func_eval_time_limit_secs=50
)

# Calculate test accuracy
y_pred = api.predict(X_test)
score = api.score(y_pred, y_test)
print("Accuracy score", score)
```

For Time Series Forecasting Tasks
```py

from autoPyTorch.api.time_series_forecasting import TimeSeriesForecastingTask

# data and metric imports
from sktime.datasets import load_longley
targets, features = load_longley()

# define the forecasting horizon
forecasting_horizon = 3

# Dataset optimized by APT-TS can be a list of np.ndarray/ pd.DataFrame where each series represents an element in the 
# list, or a single pd.DataFrame that records the series
# index information: to which series the timestep belongs? This id can be stored as the DataFrame's index or a separate
# column
# Within each series, we take the last forecasting_horizon as test targets. The items before that as training targets
# Normally the value to be forecasted should follow the training sets
y_train = [targets[: -forecasting_horizon]]
y_test = [targets[-forecasting_horizon:]]

# same for features. For uni-variant models, X_train, X_test can be omitted and set as None
X_train = [features[: -forecasting_horizon]]
# Here x_test indicates the 'known future features': they are the features known previously, features that are unknown
# could be replaced with NAN or zeros (which will not be used by our networks). If no feature is known beforehand,
# we could also omit X_test
known_future_features = list(features.columns)
X_test = [features[-forecasting_horizon:]]

start_times = [targets.index.to_timestamp()[0]]
freq = '1Y'

# initialise Auto-PyTorch api
api = TimeSeriesForecastingTask()

# Search for an ensemble of machine learning algorithms
api.search(
    X_train=X_train,
    y_train=y_train,
    X_test=X_test, 
    optimize_metric='mean_MAPE_forecasting',
    n_prediction_steps=forecasting_horizon,
    memory_limit=16 * 1024,  # Currently, forecasting models use much more memories
    freq=freq,
    start_times=start_times,
    func_eval_time_limit_secs=50,
    total_walltime_limit=60,
    min_num_test_instances=1000,  # proxy validation sets. This only works for the tasks with more than 1000 series
    known_future_features=known_future_features,
)

# our dataset could directly generate sequences for new datasets
test_sets = api.dataset.generate_test_seqs()

# Calculate test accuracy
y_pred = api.predict(test_sets)
score = api.score(y_pred, y_test)
print("Forecasting score", score)
```

For more examples including customising the search space, parellising the code, etc, checkout the `examples` folder

```sh
$ cd examples/
```


Code for the [paper](https://arxiv.org/abs/2006.13799) is available under `examples/ensemble` in the [TPAMI.2021.3067763](https://github.com/automl/Auto-PyTorch/tree/TPAMI.2021.3067763`) branch.

## Contributing

If you want to contribute to Auto-PyTorch, clone the repository and checkout our current development branch

```sh
$ git checkout development
```

### Community Contribution
1. Fork the Repository into your account
2. create a `feature` or `bug` sub-branch e.g. `git checkout -b feature/name_of_feature`
3. Create reasonable commits in your branch
4. Create a Pull Request

## License

This program is free software: you can redistribute it and/or modify
it under the terms of the Apache license 2.0 (please see the LICENSE file).

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

You should have received a copy of the Apache license 2.0
along with this program (see LICENSE file).

## Reference

Please refer to the branch `TPAMI.2021.3067763` to reproduce the paper *Auto-PyTorch Tabular: Multi-Fidelity MetaLearning for Efficient and Robust AutoDL*.

```bibtex
  @article{zimmer-tpami21a,
  author = {Lucas Zimmer and Marius Lindauer and Frank Hutter},
  title = {Auto-PyTorch Tabular: Multi-Fidelity MetaLearning for Efficient and Robust AutoDL},
  journal = {IEEE Transactions on Pattern Analysis and Machine Intelligence},
  year = {2021},
  note = {also available under https://arxiv.org/abs/2006.13799},
  pages = {3079 - 3090}
}
```

```bibtex
@incollection{mendoza-automlbook18a,
  author    = {Hector Mendoza and Aaron Klein and Matthias Feurer and Jost Tobias Springenberg and Matthias Urban and Michael Burkart and Max Dippel and Marius Lindauer and Frank Hutter},
  title     = {Towards Automatically-Tuned Deep Neural Networks},
  year      = {2018},
  month     = dec,
  editor    = {Hutter, Frank and Kotthoff, Lars and Vanschoren, Joaquin},
  booktitle = {AutoML: Methods, Sytems, Challenges},
  publisher = {Springer},
  chapter   = {7},
  pages     = {141--156}
}
```

```bibtex
@article{deng-ecml22,
  author    = {Difan Deng and Florian Karl and Frank Hutter and Bernd Bischl and Marius Lindauer},
  title     = {Efficient Automated Deep Learning for Time Series Forecasting},
  year      = {2022},
  booktitle = {Machine Learning and Knowledge Discovery in Databases. Research Track
               - European Conference, {ECML} {PKDD} 2022},
  url       = {https://doi.org/10.48550/arXiv.2205.05511},
}
```

## Contact

Auto-PyTorch is developed by the [AutoML Groups of the University of Freiburg and Hannover](http://www.automl.org/).


docs/releases.rst
:orphan:

.. _releases:

..
    The following command allows to retrieve all commiters since a specified
    commit. From http://stackoverflow.com/questions/6482436/list-of-authors-in-git-since-a-given-commit
    git log 2e29eba.. --format="%aN <%aE>" --reverse | perl -e 'my %dedupe; while (<STDIN>) { print unless $dedupe{$_}++}'


========
Releases
========

Version 0.2.1
=============
| [FIX] ADD forecasting init design to pip data files by @dengdifan in https://github.com/automl/Auto-PyTorch/pull/459
| checks for time series dataset split by @dengdifan in https://github.com/automl/Auto-PyTorch/pull/464
| [FIX] Numerical stability scaling for timeseries forecasting tasks by @dengdifan in https://github.com/automl/Auto-PyTorch/pull/467
| [FIX] pipeline options in `fit_pipeline` by @ravinkohli in https://github.com/automl/Auto-PyTorch/pull/466
| [FIX] results management and visualisation with missing test data by @ravinkohli in https://github.com/automl/Auto-PyTorch/pull/465
| [ADD] Robustly refit models in final ensemble in parallel by @ravinkohli in https://github.com/automl/Auto-PyTorch/pull/471

Contributors v0.2
*****************
* Difan Deng
* Ravin Kohli
* Shuhei Watanabe
* Theodoros Athanasiadis

Version 0.2
===========
| [FIX] Documentation and docker workflow file (#449)
| [RELEASE] Changes for release v0.2 (#446)
| [ADD] Allow users to pass feat types to tabular validator (#441)
| [ADD] docs for forecasting task (#443)
| [FIX] fit updates in gluonts (#445)
| [ADD] Time series forecasting (#434)
| [FIX] fix dist twine check for github (#439)
| [ADD] Subsampling Dataset (#398)
| [feat] Add __str__ to autoPyTorchEnum (#405)
| [ADD] feature preprocessors from autosklearn (#378)
| [refactor] Fix SparseMatrixType --> spmatrix and add ispandas (#397)
| [ADD] dataset compression (#387)
| [fix] Update the SMAC version (#388)
| [feat] Add new task inference for APT (#386)
| [FIX] Datamanager in memory (#382)
| [FIX] Fix: keyword arguments to submit (#384)
| [feat] Add coalescer (#376)
| [FIX] Remove redundant categorical imputation (#375)
| [ADD] scalers from autosklearn (#372)
| [ADD] variance thresholding (#373)
| [fix] Change int to np.int32 for the ndarray dtype specification (#371)
| [fix] Hotfix debug no training in simple intensifier (#370)
| [ADD] Test evaluator (#368)
| [FIX] Fix 361 (#367)
| [FIX] fix error after merge
| [ADD] Docker publish workflow (#357)
| [ADD] fit pipeline honoring API constraints with tests (#348)
| [FIX] Update workflow files (#363)
| [feat] Add the option to save a figure in plot setting params (#351)
| [FIX] Cleanup of simple_imputer (#346)
| [feat] Add an object that realizes the perf over time viz (#331)

Contributors v0.2
*****************

* Ravin Kohli
* Shuhei Watanabe
* Eddie Bergman
* Difan Deng

Version 0.1.1
==============
[refactor] Completely refactored version with a new scikit-learn compatible API.

Contributors v0.1.1
********************

* Ravin Kohli
* Shuhei Watanabe
* Francisco Rivera

run_bench.py
import argparse
import numpy as np
import os as os
import openml
import json
import time
import random
import torch
from sklearn.model_selection import train_test_split

import ConfigSpace as cs
from autoPyTorch import AutoNetClassification
from autoPyTorch import HyperparameterSearchSpaceUpdates
from autoPyTorch.pipeline.nodes import LogFunctionsSelector
from autoPyTorch.components.metrics.additional_logs import *


def resplit(X, y):
    test_split = 0.33
    indices = np.array(range(len(y)))
    ind_train, ind_test = train_test_split(indices, stratify=y, test_size=test_split, shuffle=True)
    return ind_train, ind_test

def get_sampling_space():
    sampling_space = dict()
    sampling_space["batch_loss_computation_techniques"] = ['standard', 'mixup']
    sampling_space["networks"] = ['shapedmlpnet']
    sampling_space["optimizer"] = ['adam', 'sgd', 'adamw', 'rmsprop']
    sampling_space["preprocessors"] = ['truncated_svd', 'none']
    sampling_space["imputation_strategies"] = ['mean']
    sampling_space["lr_scheduler"] = ['cosine_annealing', 'step']
    sampling_space["normalization_strategies"] = ['standardize']
    sampling_space["over_sampling_methods"] = ['none', 'random', 'smote']
    sampling_space["under_sampling_methods"] = ['none', 'random']
    sampling_space["target_size_strategies"] = ['upsample', 'downsample']
    sampling_space["embeddings"] = ['learned']
    sampling_space["initialization_methods"] = ['default', 'sparse']
    sampling_space["loss_modules"] = ['cross_entropy_weighted']
    sampling_space["best_over_epochs"] = False
    sampling_space["budget_type"] = "epochs"
    sampling_space["refit_validation_split"] = 0.33
    sampling_space["validation_split"] = 0.33
    sampling_space["log_level"] = "info"
    sampling_space["task_id"] = 0
    sampling_space["use_tensorboard_logger"] = True
    sampling_space["full_eval_each_epoch"] = True
    sampling_space["log_every_n_datapoints"] = None
    sampling_space["optimize_metric"] = "accuracy"
    sampling_space["additional_metrics"] = ["cross_entropy", "balanced_accuracy"]
    return sampling_space


def load_hyperparameter_config(config_dir):
    with open(config_dir, "r") as f:
        data = json.load(f)
    for key,val in data.items():
        if val=="True":
            data[key] = True
        elif val=="False":
            data[key] = False
    return data


def seed_everything(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)


def get_openml_task_ids():
    openml_task_ids = [167149, 167152, 167161, 167168, 167181, 126025, 167190, 126026, 167185, # 0.3 split (23)
                       167184, 126029, 167201, 189905, 189906, 189909, 167083, 189908, 167104,
                       189865, 189862, 189866, 189873, 167200,
                       189354, 3945, 7593, 146212, 34539, 168331, 168330, 168335, 168329, 168868, #0.1 split (16)
                       168908, 168910]

    resplit_tasks = [189354, 3945, 7593, 146212, 34539, 168332, 168331, 168330, 168337, 168335,
                     168338, 168329, 168868, 168908, 168910, 168909]
    
    return openml_task_ids, resplit_tasks


def load_openml_data(openml_task_id, resplit_tasks):
    task = openml.tasks.get_task(task_id=openml_task_id)
    X, y = task.get_X_and_y()
    ind_train, ind_test = task.get_train_test_split_indices()

    if openml_task_id in resplit_tasks:
       ind_train, ind_test = resplit(X, y)
    
    return X, y, ind_train, ind_test

def get_search_space_updates():
    search_space_updates = HyperparameterSearchSpaceUpdates()
    search_space_updates.append(node_name="LearningrateSchedulerSelector",
                                hyperparameter="cosine_annealing:T_max",
                                value_range=[budget, budget])
    search_space_updates.append(node_name="NetworkSelector",
                                hyperparameter="shapedmlpnet:max_units",
                                value_range=[64, 1024],
                                log=True)
    search_space_updates.append(node_name="NetworkSelector",
                                hyperparameter="shapedmlpnet:num_layers",
                                value_range=[2, 8])
    search_space_updates.append(node_name="NetworkSelector",
                                hyperparameter="shapedresnet:max_units",
                                value_range=[10,1024])
    search_space_updates.append(node_name="NetworkSelector",
                                hyperparameter="shapedresnet:num_groups",
                                value_range=[1,9])
    search_space_updates.append(node_name="NetworkSelector",
                                hyperparameter="shapedresnet:blocks_per_group",
                                value_range=[1,4])
    return search_space_updates


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Fit a random config on an openml task')
    parser.add_argument("--run_id", type=int, help="An id for the run.")
    parser.add_argument("--offset", type=int, help="An offset to get larger array numbers than allowed by the cluster", default=0)
    parser.add_argument("--seed", type=int)
    parser.add_argument("--budget", type=int)
    parser.add_argument("--config_dir", type=str)
    parser.add_argument("--root_logdir", type=str)
    args = parser.parse_args()

    # args
    run_id = args.run_id + args.offset
    seed = args.seed
    budget = args.budget
    root_logdir = args.root_logdir
    config_dir = args.config_dir

    # Data splitting seed
    seed_everything(42)

    # Get data
    openml_task_ids, resplit_tasks = get_openml_task_ids()
    openml_task_id = openml_task_ids[run_id % len(openml_task_ids)]
    X, y, ind_train, ind_test = load_openml_data(openml_task_id, resplit_tasks)

    # Settings
    logdir = os.path.join(root_logdir, "budget_"+str(budget), "seed_"+str(seed), "run_" + str(run_id) + "_" + str(openml_task_id))
    hyperparameter_config_dir = os.path.join(config_dir, "config_"+ str(run_id//len(openml_task_ids)) + ".json")

    print("Using config", str(run_id//len(openml_task_ids)), "for task", openml_task_id, "and run id", run_id)

    # Training seed
    seed_everything(seed)

    # Sample config (autonet)
    search_space_updates = get_search_space_updates()
    sampling_space = get_sampling_space()
    sampling_space["random_seed"] = seed
    sampling_space["run_id"] = run_id
    sampling_space["result_logger_dir"] = logdir
    sampling_space["additional_logs"] = [test_result.__name__, test_cross_entropy.__name__, test_balanced_accuracy.__name__]

    # Initialize Autonet
    autonet = AutoNetClassification(**sampling_space,
                                    hyperparameter_search_space_updates=search_space_updates)

    # Add additional logs
    gl = GradientLogger()
    lw_gl = LayerWiseGradientLogger()
    additional_logs = [gradient_max(gl), gradient_mean(gl), gradient_median(gl), gradient_std(gl),
                       gradient_q10(gl), gradient_q25(gl), gradient_q75(gl), gradient_q90(gl),
                       layer_wise_gradient_max(lw_gl), layer_wise_gradient_mean(lw_gl),
                       layer_wise_gradient_median(lw_gl), layer_wise_gradient_std(lw_gl),
                       layer_wise_gradient_q10(lw_gl), layer_wise_gradient_q25(lw_gl),
                       layer_wise_gradient_q75(lw_gl), layer_wise_gradient_q90(lw_gl),
                       gradient_norm()]

    for additional_log in additional_logs:
        autonet.pipeline[LogFunctionsSelector.get_name()].add_log_function(name=type(additional_log).__name__,
                                                                           log_function=additional_log)

        sampling_space["additional_logs"].append(type(additional_log).__name__)

    autonet.pipeline[LogFunctionsSelector.get_name()].add_log_function(name=test_result.__name__, 
                                                                       log_function=test_result(autonet, X[ind_test], y[ind_test]))
    autonet.pipeline[LogFunctionsSelector.get_name()].add_log_function(name=test_cross_entropy.__name__,
                                                                       log_function=test_cross_entropy(autonet, X[ind_test], y[ind_test]))
    autonet.pipeline[LogFunctionsSelector.get_name()].add_log_function(name=test_balanced_accuracy.__name__,
                                                                       log_function=test_balanced_accuracy(autonet, X[ind_test], y[ind_test]))


    # Load hyperparameters
    hyperparameter_config = load_hyperparameter_config(hyperparameter_config_dir)
    if "LearningrateSchedulerSelector:cosine_annealing:T_max" in hyperparameter_config.keys():
        hyperparameter_config["LearningrateSchedulerSelector:cosine_annealing:T_max"] = budget

    # Print Infos
    info = {
            "OpenML_task_id" : int(openml_task_id),
            "test_split" : len(y[ind_test])/(len(y[ind_train])+len(y[ind_test])),
            "budget": budget,
            "seed" : int(seed),
            "instances" : int(len(y[ind_train])+len(y[ind_test])),
            "classes": int(len(np.unique(y[ind_train]))),
            "features": int(X.shape[1])
            }

    print("Autonet config:", autonet.get_current_autonet_config())
    print("Hyperparameter config:", hyperparameter_config)
    print("Info", info)

    # Refit
    results = autonet.refit(X_train=X[ind_train],
                            Y_train=y[ind_train],
                            hyperparameter_config=hyperparameter_config,
                            autonet_config=autonet.get_current_autonet_config(),
                            budget=budget)

    # Write to json
    results["openml_task_id"] = int(openml_task_id)
    results["run_id"] = int(run_id)
    results["validation_split"] = sampling_space["refit_validation_split"]

    pop_keys = []
    for key in results["info"].keys():
        if key.startswith("layer"):
            pop_keys.append(key)

    for key in pop_keys:
        del results["info"][key]

    with open(logdir + "/results_dump.json", "w") as f:
        json.dump(results, f)

    with open(logdir + "/info.json", "w") as f:
        json.dump(info, f)
